src/common/mem.c

```diff
@@ -1573,4 +1573,362 @@ umem_tx_publish(struct umem_instance *umm, struct umem_rsrvd_act *rsrvd_act)
+	}
+	sgl->sg_nr_out = sgl->sg_nr = nr;
+	store_iod->io_nr            = nr;
+}
+
+int
+umem_cache_checkpoint(struct umem_cache *cache, uint64_t wal_tx)
+{
+	struct umem_store           *store    = cache->ca_store;
+	struct umem_page            *page     = NULL;
+	struct umem_page            *tmp      = NULL;
+	struct umem_checkpoint_data *chkpt_data_all;
+	struct umem_checkpoint_data *chkpt_data;
+	uint64_t                     committed_tx;
+	d_list_t                     free_list;
+	int                          i;
+	int                          rc;
+	int                          inflight = 0;
+
+	if (d_list_empty(&cache->ca_pgs_dirty))
+		return 0;
+
+	D_ASSERT(store != NULL);
+
+	D_INIT_LIST_HEAD(&free_list);
+	D_ALLOC_ARRAY(chkpt_data_all, MAX_INFLIGHT);
+	if (chkpt_data_all == NULL)
+		return -DER_NOMEM;
+
+	/** Setup the inflight IODs */
+	for (i = 0; i < MAX_INFLIGHT; i++) {
+		chkpt_data = &chkpt_data_all[i];
+		d_list_add_tail(&chkpt_data->cd_link, &free_list);
+		chkpt_data->cd_store_iod.io_regions = &chkpt_data->cd_regions[0];
+		chkpt_data->cd_sg_list.sg_iovs      = &chkpt_data->cd_iovs[0];
+	}
+
+	d_list_splice_init(&cache->ca_pgs_dirty, &cache->ca_pgs_copying);
+
+	/** First mark all pages in the new list so they won't be moved by an I/O thread.  This
+	 *  will enable us to continue the algorithm in relative isolation from I/O threads.
+	 */
+	d_list_for_each_entry(page, &cache->ca_pgs_copying, pg_link) {
+		/** Mark all pages in copying list first.  Marking them as waiting will prevent
+		 *  them from being moved to another list by an I/O operation.
+		 */
+		page->pg_waiting = 1;
+	}
+
+	do {
+		/** first try to add up to MAX_INFLIGHT pages to the waiting queue */
+		while (inflight < MAX_INFLIGHT &&
+		       (page = d_list_pop_entry(&cache->ca_pgs_copying, struct umem_page,
+						pg_link)) != NULL) {
+			chkpt_data =
+			    d_list_pop_entry(&free_list, struct umem_checkpoint_data, cd_link);
+			D_ASSERT(chkpt_data != NULL);
+			page2chkpt(page, chkpt_data);
+
+			rc = store->stor_ops->so_flush_prep(store, &chkpt_data->cd_store_iod,
+							    &chkpt_data->cd_fh);
+
+			/** Need to figure out what errors are possible here and how to handle
+			 *  them.  This is wrong but will suffice to get the system kicking
+			 */
+			D_ASSERT(rc == 0);
+
+			page->pg_copying         = 1;
+			page->pg_last_checkpoint = page->pg_last_inflight;
+			page->pg_dirty           = 0;
+
+			rc = store->stor_ops->so_flush_copy(chkpt_data->cd_fh,
+							    &chkpt_data->cd_sg_list);
+			/** Same comment as above.  Fix this later */
+			D_ASSERT(rc == 0);
+
+			page->pg_copying = 0;
+
+			d_list_add_tail(&page->pg_link, &cache->ca_pgs_waiting);
+
+			memset(&page->pg_bmap[0], 0, sizeof(page->pg_bmap));
+			inflight++;
+		}
+
+		store->stor_ops->so_wal_query_committed(store, &committed_tx);
+
+		d_list_for_each_entry_safe(page, tmp, &cache->ca_pgs_waiting, pg_link) {
+			if (page->pg_last_checkpoint > committed_tx)
```

Comment:
We have an ID compare callback provided, see the comment in wal_id_cmp(), it tries to handle sequence overflow case, which is very unlikely though.

Reply:
Will fix this